courses/C06-pkg_avance_test.Rmd

```diff
@@ -0,0 +1,247 @@
+test_that("test erreur save_as_csv", {
+  data(iris)
+  expect_error(save_as_csv(iris, "nexiste/pas/out.csv"), "does not exist")
+  expect_error(save_as_csv(iris, "fichier.R"), "does not have extension")
+})
+
+test_that("test conformite save_as_csv", {
+  data(iris)
+  expect_is(save_as_csv(iris, "out.csv"), "character")
+  expect_true(file.exists(save_as_csv(iris, "out.csv")))
+  expect_equal(iris,read.csv2("out.csv"))
+})
+
+```
+
+???
+
+Nice...
+
+
+---
+class: slide 
+
+### Automate the testing of your code
+ 
+Launch tests:
+
+```{r eval=FALSE}
+devtools::test()
+```
+
+
+```{r, echo=FALSE, out.width="35%"}
+knitr::include_graphics("images/exemple_test.png")
+```
+
+
++ Each line corresponds to a test file
+
++ Each test is counted as `OK`, `FAIL`, `WARNING`, `SKIP`
+
++ Each test is launched in its own environment, it is necessary to load the necessary packages and data 
+
+???
+
+
+To talk a little about {fusen}. In the thumbnail, before sending all the pieces of code to the right place
+
+Separate the function and its roxygen in a chunk named function-1
+
+Separate the usage examples in a chunk named examples-1
+
+Separate the unit tests, with the complete test_that("", {}) structure, in a chunk named tests-1
+
+Indicate that at this point, if this Rmd was in the "dev/" folder, the {fusen} package would move everything to where it needs to be, without anything else to do.
+
+---
+class: slide
+
+### Quiz: How to test what is supposed to fail?
+
+- A: `expect_equal()`
+- B: `expect_false()`
+- C: `expect_error()`
+- D: `expect_warning()`
+
+???
+
+- If your function crashes, R returns an Error. 
+You can catch it with `expect_error()` to avoid crashing the test script
+and check that in this case there is an error.  
+It is best if the error is chosen and made explicit in your function with a `stop()` for example
+
+- Depending on your definition of fail, `expect_warning()` can also be a good answer. 
+Generally speaking, a warning should be something that the user has to correct in his code. It is good practice to put `stop()` or `message()`, but not `warning()` in your functions.
+
+---
+class: slide 
+exclude: true
+
+### TD 
+
+**Version A**
+
+Add unit tests to your {tools} package containing the `save_as_csv()` and `clean_names()` functions.
+
+Your package must be documented, **tested** and return 0 errors, 0 warnings, and 0 notes
+
+**Bonus**
+
+Add to your {tools} package, working in a new thumbnail:
+
++ A function that returns the first 6 values of a chosen column (as a parameter) from a dataset. The output is a vector.
+    + `extract_six_from_col(mtcars, "mpg")` works
+    + `extract_six_from_col(mtcars, "toto")` returns an error
++ unit tests for `extract_six_from_col()`
+
+Your package must be documented, **tested** and return 0 errors, 0 warnings, and 0 notes
+
+---
+class: slide
+
+### Hands-on
+
+- Add unit test in your {hello} package to test that `say_hello()` returns a message
+- Check the package
+
```

Comment:
I wonder if we could/should add anything about how to come up with good test cases? I often think new developers do not think carefully about creating _hypotheses_ for how their code could fail and evaluating actual corner cases